internal/index/scip.go

```diff
@@ -0,0 +1,432 @@
+		if module, ok := dependencies[moduleName]; ok {
+			return &module
+		}
+	}
+
+	return nil
+}
+
+func emitImportMoniker(dependencies map[string]gomod.GoModule, obj ObjectLike) bool {
+	// pkg := makeMonikerPackage(obj)
+	// monikerIdentifier := joinMonikerParts(pkg, makeMonikerIdentifier(i.packageDataCache, p, obj))
+
+	// for _, moduleName := range packagePrefixes(pkg) {
+	// 	if module, ok := dependencies[moduleName]; ok {
+	// 		// Lazily emit package information vertex
+	// 		packageInformationID := i.ensurePackageInformation(module.Name, module.Version)
+	//
+	// 		// Lazily emit moniker vertex
+	// 		monikerID := i.ensureImportMoniker(monikerIdentifier, packageInformationID)
+	//
+	// 		// Monikers will be linked during Indexer.linkImportMonikersToRanges
+	// 		i.addImportMonikerReference(monikerID, rangeID, document.DocumentID)
+	//
+	// 		return true
+	// 	}
+	// }
+
+	return false
+}
+
+// packagePrefixes returns all prefix of the go package path. For example, the package
+// `foo/bar/baz` will return the slice containing `foo/bar/baz`, `foo/bar`, and `foo`.
+func packagePrefixes(packageName string) []string {
+	parts := strings.Split(packageName, "/")
+	prefixes := make([]string, len(parts))
+
+	for i := 1; i <= len(parts); i++ {
+		prefixes[len(parts)-i] = strings.Join(parts[:i], "/")
+	}
+
+	return prefixes
+}
+
+// A Visitor's Visit method is invoked for each node encountered by Walk.
+// If the result visitor w is not nil, Walk visits each of the children
+// of node with the visitor w, followed by a call of w.Visit(nil).
+// type Visitor interface { Visit(node Node) (w Visitor) }
+
+type FileVisitor struct {
+	deps map[string]gomod.GoModule
+	pkg  *PackageInfo
+	doc  *scip.Document
+	file *ast.File
+}
+
+func (f FileVisitor) Visit(n ast.Node) (w ast.Visitor) {
+	if n == nil {
+		return nil
+	}
+
+	switch node := n.(type) {
+	case *ast.File:
+		// explicit pass
+	case *ast.FuncDecl:
+		// node.Name
+		fmt.Println(node.Name.Name)
+	case *ast.Ident:
+		fmt.Println("=======")
+		info := f.pkg.TypesInfo
+
+		def := info.Defs[node]
+		ref := info.Uses[node]
+
+		// Don't think anything can be a def and a ref
+		if def != nil && ref != nil {
+			panic("Didn't think this was possible")
+		}
+
+		pos := node.NamePos
+		position := f.pkg.Fset.Position(pos)
+
+		// Append definition
+		if def != nil {
+			f.doc.Occurrences = append(f.doc.Occurrences, &scip.Occurrence{
+				Range:       scipRange(position, def),
+				Symbol:      scipSymbol(findModuleForObj(f.deps, def), def),
+				SymbolRoles: int32(scip.SymbolRole_Definition),
+			})
+		} else if ref != nil {
+			f.doc.Occurrences = append(f.doc.Occurrences, &scip.Occurrence{
+				Range:       scipRange(position, ref),
+				Symbol:      scipSymbol(findModuleForObj(f.deps, ref), ref),
+				SymbolRoles: int32(scip.SymbolRole_ReadAccess),
+			})
+		}
+
+		fmt.Println("Definition: ", def)
+		fmt.Println("References: ", ref)
+	default:
+		fmt.Printf("unhandled: %T %v\n", n, n)
```

Comment:
Is there a way to fallback to the default visitor that continues recursing on the AST?

Reply:
well it will continue to recurse the AST even with this strategy -- I just am letting myself know that we didn't handle something